### ISSUE TITLE
Forbidden filenames on Windows exists in git index will fail parse_index()

### ISSUE THREAD

**Issue by ivellioscolin** [libgit2/libgit2#3588]
Well this is regarding finding root cause of #3585 .
At first I suspect it's related to remote folder access since I never see it in local repository on my Windows drive. But it's not. It's related to forbidden filename on Windows. **If a git entry contains a file with forbidden name on Windows, the issue will show up.**

Reproduce:
1. In Linux environment, git init a new repository. Add a new file named "aux" or any other forbidden file name for Windows. Commit it to add to git repository. Copy the entire repository folder to a Windows environment, or just add it for network sharing to be remotely accessed.
2. In Windows environment, since such file name is forbidden, the file can't be copied. If you access it from network, the file name will be automatically changed to another by OS.
3. Install Github's Atom. Open any file from the repository. 
>You will see git.node crashes.

Details:
git_index_open()
>git_index_read()
>>parse_index()
>>>read_entry()    <<<< When loop all the entries, copy the entry from buffer which contains the invalid name.
>>>>...
>>>>>verify_component() <<< When verifying the dospath, any invalid name like "CON", "AUX", etc will fail the verification. Then return a entry with entry_size==0. Then the entire parse_index() will abort.

git_iterator_free() will be called when git iterator tree fails. Then it hits the NULL pointer which #3585 trying to fix.

Can we simply consider such entry as "corrupted"? May need a solution for the issue.

**Comment by ivellioscolin** [comment]
@ethomson

**Comment by tkelman** [comment]
I see a similar failure of libgit2 to clone a repo when any of its tags have quotes in them, e.g. https://github.com/tkelman/Example.jl/releases

```
#include "git2.h"
void main() {
    git_repository* repo_ptr = NULL;
    char* repo_url = "https://github.com/tkelman/Example.jl"; // fails, change tkelman to JuliaLang (no illegal tag names) and it works
    char* repo_path = "Example.Bare";
    git_clone_options clone_opts = GIT_CLONE_OPTIONS_INIT;
    clone_opts.bare = 1;
    git_libgit2_init();
    git_clone(&repo_ptr, repo_url, repo_path, &clone_opts);
    git_libgit2_shutdown();
}
```

**Comment by babzog** [comment]
Another reproduction scenario: Create a file in Linux with a dot "." at the end, save and commit.  If you check out the repo from within Windows, not only will it complain about that file, but, it will be marked as deleted, forever making your repo a pain to use.  Further, this issue arises, whereby all of the icon overlays for the entire code tree are hosed, owing to this one bad entry.

The lib should not fail because of an invalid entry, it should simply be skipped.

**Comment by ethomson** [comment]
@babzog:  I'm not entirely sure that I understand.  What you're describing seems to be limitations in Tortoise Git - though I'm sort of guessing that based on the fact that you mention icon overlays, which is a Windows Shell terminology.

The behavior that you're describing (invalid file paths appearing as deleted) is behavior in *git itself*, where it *does* skip the invalid entry.  Since it skips the entry, it isn't written to disk, so it is shown as deleted, because it could not possibly be shown as anything else, since the entry is in the index but is not (and cannot be, by definition) on disk.  (Though, from my testing, this behavior seems to have changed in more recent versions of Git for Windows.)

In any case, if you have complaints with Tortoise Git, I would urge you to open an issue with them.

**Comment by ethomson** [comment]
Fixed via #3619

**Comment by ethomson** [comment]
@tkelman Thanks for the report - I split your issue into #3622
